src/sdam/monitor.rs

```diff
@@ -1,82 +1,132 @@
+    update_receiver: ServerUpdateReceiver,
+}
+
+impl Monitor {
+    /// Creates a monitor associated with a given Server.
+    /// This method does not start the monitor. Use `Monitor::start` to do so.
+    pub(super) fn new(
+        address: StreamAddress,
+        server: &Arc<Server>,
+        client_options: ClientOptions,
+        update_receiver: ServerUpdateReceiver,
+    ) -> Self {
+        Self {
+            address,
+            server: server.clone(),
+            client_options,
+            update_receiver,
+        }
+    }
+
+    /// Start the monitor tasks.
+    /// A weak reference is used to ensure that the monitor doesn't keep the topology alive after
+    /// it's been removed from the topology or the client has been dropped.
+    pub(super) fn start(self, topology: WeakTopology) {
+        let mut heartbeat_monitor = HeartbeatMonitor::new(
+            self.address,
+            Arc::downgrade(&self.server),
+            self.client_options,
+        );
+        let t = topology.clone();
+        RUNTIME.execute(async move {
+            heartbeat_monitor.execute(t).await;
+        });
+
+        let update_monitor = UpdateMonitor {
+            server: Arc::downgrade(&self.server),
+            update_receiver: self.update_receiver,
+            generation_subscriber: self.server.pool.subscribe_to_generation_updates(),
+        };
+        RUNTIME.execute(async move {
+            update_monitor.execute(topology).await;
+        });
+    }
+}
+
+/// Monitor that performs regular heartbeats to determine server status.
+struct HeartbeatMonitor {
     address: StreamAddress,
     connection: Option<Connection>,
     handshaker: Handshaker,
     server: Weak<Server>,
-    server_type: ServerType,
-    topology: WeakTopology,
+    client_options: ClientOptions,
 }
 
-impl Monitor {
-    /// Starts a monitoring thread associated with a given Server. A weak reference is used to
-    /// ensure that the monitoring thread doesn't keep the server alive after it's been removed
-    /// from the topology or the client has been dropped.
-    pub(super) fn start(address: StreamAddress, server: Weak<Server>, topology: WeakTopology) {
-        let handshaker = Handshaker::new(Some(topology.client_options().clone().into()));
-        let mut monitor = Self {
+impl HeartbeatMonitor {
+    fn new(address: StreamAddress, server: Weak<Server>, client_options: ClientOptions) -> Self {
+        let handshaker = Handshaker::new(Some(client_options.clone().into()));
+        Self {
             address,
-            connection: None,
             server,
-            server_type: ServerType::Unknown,
-            topology,
+            client_options,
             handshaker,
-        };
-
-        RUNTIME.execute(async move {
-            monitor.execute().await;
-        });
+            connection: None,
+        }
     }
 
-    async fn execute(&mut self) {
+    async fn execute(&mut self, weak_topology: WeakTopology) {
```

Comment:
Out of curiosity, what's the rationale for having the topology be a parameter to execute instead of a field on the struct? (Likewise for UpdateMonitor below)

Reply:
I moved the creation of the `Monitor` struct to happen at the same time as the creation of the `Server`, but when a `Topology` is being created itself it makes a bunch of `Server`s first to avoid having to access a lock and make `Topology::new` async. Now that I think about it though, we can use `futures::executor::block_on` like we do for `watch` channels since that lock would be acquired immediately, and we can avoid this situation altogether. Should I go ahead and make that change?

Reply:
Yep, that sounds good to mean, provided that there's a comment indicating why `block_on` is fine similar to the one for the channels.

Reply:
So it looks like the `Topology::new_from_hosts` also needs the ability to create a server without starting a monitor, so I'm thinking it's probably best if we just leave it as it is.

Reply:
`new_from_hosts` is just used for tests IIRC, so we should feel free to work around if (e.g. write a new method or something) if needed.

Reply:
Updated to store the `WeakTopology` on the monitor. I kept the two phase `execute` (which now takes no parameters) to allow for `new_from_hosts` to create servers without starting the monitors.